pkg/search/postgres/storecache.go

```diff
@@ -0,0 +1,329 @@
+package postgres
+
+import (
+	"context"
+
+	v1 "github.com/stackrox/rox/generated/api/v1"
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/auth/permissions"
+	"github.com/stackrox/rox/pkg/postgres"
+	"github.com/stackrox/rox/pkg/postgres/walker"
+	"github.com/stackrox/rox/pkg/sac"
+	"github.com/stackrox/rox/pkg/sync"
+)
+
+// NewGenericStoreWithCache returns new subStore implementation for given resource.
+// subStore implements subset of Store operations.
+func NewGenericStoreWithCache[T any, PT unmarshaler[T]](
+	db postgres.DB,
+	schema *walker.Schema,
+	pkGetter primaryKeyGetter[T, PT],
+	insertInto inserter[T, PT],
+	copyFromObj copier[T, PT],
+	setAcquireDBConnDuration durationTimeSetter,
+	setPostgresOperationDurationTime durationTimeSetter,
+	upsertAllowed upsertChecker[T, PT],
+	targetResource permissions.ResourceMetadata,
+) Store[T, PT] {
+	underlyingStore := NewGenericStore[T, PT](
+		db,
+		schema,
+		pkGetter,
+		insertInto,
+		copyFromObj,
+		setAcquireDBConnDuration,
+		setPostgresOperationDurationTime,
+		upsertAllowed,
+		targetResource,
+	)
+	store := &CachedStore[T, PT]{
+		schema:          schema,
+		pkGetter:        pkGetter,
+		targetResource:  targetResource,
+		cache:           make(map[string]PT),
+		useCache:        true,
+		underlyingStore: underlyingStore,
+	}
+	store.cacheLock.Lock()
+	defer store.cacheLock.Unlock()
+	store.resetCacheNoLock()
```

Comment:
nit: maybe a comment clarifying that the first "reset" call is actually populating the cache with store values, just reading through this may seem unexpected.